packages/eslint-plugin/src/rules/strict-comparisons.ts

```diff
@@ -0,0 +1,310 @@
+import { TSESTree } from '@typescript-eslint/typescript-estree';
+import { isUnionType, isTypeFlagSet } from 'tsutils';
+import ts from 'typescript';
+
+import * as util from '../util';
+
+interface Config {
+  allowObjectEqualComparison: boolean;
+  allowStringOrderComparison: boolean;
+}
+
+type Options = [Config];
+
+type MessageIds = 'nonComparableTypes' | 'invalidTypeForOperator';
+
+export default util.createRule<Options, MessageIds>({
+  name: 'strict-comparisons',
+  meta: {
+    type: 'problem',
+    docs: {
+      category: 'Best Practices',
+      description: 'Only allow comparisons between primitive types.',
+      tslintRuleName: 'object-comparison',
+      recommended: 'error',
+    },
+    messages: {
+      nonComparableTypes:
+        "cannot compare type '{{ typesLeft }}' to type '{{ typesRight }}'",
+      invalidTypeForOperator:
+        "cannot use '{{ comparator }}' comparator for type '{{ type }}'",
+    },
+    schema: [
+      {
+        type: 'object',
+        properties: {
+          allowObjectEqualComparison: {
+            type: 'boolean',
+          },
+          allowStringOrderComparison: {
+            type: 'boolean',
+          },
+        },
+        additionalProperties: false,
+      },
+    ],
+  },
+  defaultOptions: [
+    {
+      allowObjectEqualComparison: false,
+      allowStringOrderComparison: false,
+    },
+  ],
+  create(
+    context,
+    [{ allowObjectEqualComparison, allowStringOrderComparison }],
+  ) {
+    const service = util.getParserServices(context);
+
+    const typeChecker = service.program.getTypeChecker();
+
+    const enum TypeKind {
+      Any = 0,
+      Number = 1,
+      Enum = 2,
+      String = 3,
+      Boolean = 4,
+      NullOrUndefined = 5,
+      Object = 6,
+    }
+
+    const typeNames = {
+      [TypeKind.Any]: 'any',
+      [TypeKind.Number]: 'number',
+      [TypeKind.Enum]: 'enum',
+      [TypeKind.String]: 'string',
+      [TypeKind.Boolean]: 'boolean',
+      [TypeKind.NullOrUndefined]: 'null | undefined',
+      [TypeKind.Object]: 'object',
+    };
+
+    /**
+     * Get TypeKinds of a typescript type
+     * @param type
+     */
+    function getKinds(type: ts.Type): TypeKind[] {
+      return isUnionType(type)
+        ? Array.from(new Set(type.types.map(getKind)))
+        : [getKind(type)];
+    }
+
+    /**
+     * Get TypeKind of a typescript type
+     * @param type
+     */
+
+    function getKind(type: ts.Type): TypeKind {
+      // tslint:disable:no-bitwise
+      return is(ts.TypeFlags.String | ts.TypeFlags.StringLiteral)
+        ? TypeKind.String
+        : is(ts.TypeFlags.Number | ts.TypeFlags.NumberLiteral)
+        ? TypeKind.Number
+        : is(ts.TypeFlags.BooleanLike)
+        ? TypeKind.Boolean
+        : is(ts.TypeFlags.Null | ts.TypeFlags.Undefined | ts.TypeFlags.Void)
+        ? TypeKind.NullOrUndefined
+        : is(ts.TypeFlags.Any)
+        ? TypeKind.Any
+        : TypeKind.Object;
```

Comment:
I know it saves characters, but I personally am against nested ternaries because it's confusing to read.